### ISSUE TITLE
Go Producer Much slower than Python Producer

### ISSUE THREAD

**Issue by abhinavsolan** [confluentinc/confluent-kafka-go#567]
Description
===========
Producer Slowness in confluent-kafka-go

We have 2 Kafka Producers working with given settings one in Python and one in Go, Go Kafka Producer need to change the `queue.buffering.max.ms` to `50` to make it performant while a lower setting in Python is able to handle much higher load.

Even after setting the value of  `queue.buffering.max.ms` to 50 in Go, `p99` of `int_latency` for few partitions is around 150ms, while in Python it's almost negligible

Is there a reason on why the confluent-kafka-go producer is slower, or should I change few properties to align with go producer version?

Python Producer
```
request.timeout.ms = 2000
retry.backoff.ms = 500
socket.nagle.disable = True
queue.buffering.max.messages = 100000
queue.buffering.backpressure.threshold = 1
queue.buffering.max.ms = 0.5
retries=10
```
This producer is producing to a topic with 1024 partitions 
Peak overall load on Brokers - 20mb/s
Peak load on one Broker - 1.2 mb/s
librdkafka version - 1.4.2

Go Producer
```
request.timeout.ms = 2000
retry.backoff.ms = 500
socket.nagle.disable = True
queue.buffering.max.messages = 100000
queue.buffering.backpressure.threshold = 1
queue.buffering.max.ms = 50
retries=10
```
This Producer is producing to a topic with 500 partitions
Peak load on Brokers - 10 mb/s
Peak load on one Broker - 600 Kb/s
librdkafka version - 1.4.2
confluent-go - 1.4.2

Kafka Broker Version for Both - 2.4.0
OS - Linux

How to reproduce
================

Checklist
=========
Please provide the following information:

 - [X] confluent-kafka-go and librdkafka version (`LibraryVersion()`):
 - [X] Apache Kafka broker version:
 - [X] Client configuration: `ConfigMap{...}`
 - [ ] Operating system:
 - [ ] Provide client logs (with `"debug": ".."` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
Can you share your produce loop code?

**Comment by abhinavsolan** [comment]
Sure, we start a long running go routine to handle callbacks 
```
func (p *producerClient) Run(ctx context.Context) {
	defer func() {
		if r := recover(); r != nil {
			p.Log().Error(ctx, fmt.Sprintf("caught panic in producer delivery handler, %v", r))
		}
		p.Log().Info(ctx, "closing kafka producer delivery handler loop")
	}()
	for {
		select {
		case <-ctx.Done():
			return
		case e, ok := <-p.producer.Events():
			if !ok {
				return
			}
			switch ev := e.(type) {
			case *kafka.Message:
				if ev.TopicPartition.Error != nil {
					p.Log().Error(ctx, fmt.Sprintf("Kafka Produce Failed: %v\n", ev.TopicPartition))
					p.sendFailedStats(ev)
				} else {
					p.sendSuccessStats(ev)
				}
			case *kafka.Stats:
				go stats.PublishKafkaStats(p, []byte(ev.String()))
			default:
				// do nothing
			}
		}
	}
}
```

And then this the call that's been made on each message production
```
func (p *producerClient) Produce(key string, streamingLocs []*locations_v2.StreamingLocation) error {
	msg, err := createKafkaMsg(key, streamingLocs, p.topicName)
	if err != nil {
		return err
	}
	for _, m := range msg {
		p.producer.ProduceChannel() <- m
	}
	return nil
}
```

**Comment by edenhill** [comment]
What is the produce rate in messages per second?

**Comment by edenhill** [comment]
Have you tried using the non-channel-based Produce()?

**Comment by abhinavsolan** [comment]
Produce Rate for both is around 33K Messages / sec

**Comment by abhinavsolan** [comment]
Yes, we tried with non-channel-based Producer, but that was worse, it was not able to handle the load

**Comment by edenhill** [comment]
33K msgs/s does not sound like much.
What is the CPU usage of the Go application?

**Comment by abhinavsolan** [comment]
The CPU usage goes upto 25% during peak traffic, and also memory usage is around 30% only

**Comment by deankarn** [comment]
I'm also seeing this with a very low flow producer, maybe 1rps(a test application not in production yet why so low).

I have `queue.buffering.max.ms` set to `5ms` and `required acks` to `-1` with replication factor of 3(so limited not all nodes in cluster).

With that am seeing it taking `50ms->100ms` to receive the confirmation from the `producer.Events()` channel. I do see that the data does appear to be written timely and consumed immediately by another consumer, so suspecting it's somehow  the reporting to the events channel.

Using the latest `confluent-kafka-go v2.2.0`.
